packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2017 Erisson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { ContainerModule } from "inversify";
+import {
+        ConnectionHandler,
+        JsonRpcConnectionHandler,
+} from "@theia/core/lib/common";
+import {
+        SearchInWorkspaceServer,
+        SearchInWorkspaceClient,
+} from "../common/search-in-workspace-interface";
+import { RipgrepSearchInWorkspaceServer } from "./ripgrep-search-in-workspace-server";
+
+export default new ContainerModule(bind => {
+        bind(SearchInWorkspaceServer).to(RipgrepSearchInWorkspaceServer).inSingletonScope();
+        bind(ConnectionHandler).toDynamicValue(ctx =>
+                new JsonRpcConnectionHandler<SearchInWorkspaceClient>
+                        ('/search-in-workspace', client => {
+                                const server = ctx.container.get<SearchInWorkspaceServer>(SearchInWorkspaceServer);
+                                server.setClient(client);
```

Comment:
This will reset a single client, whenever a new connection is created. I.e. if you have two frontends only the last one will receive results.

Reply:
What would be the right way to do it?

Reply:
either each client should receive own instance of `SearchInWorkspaceServer` or use a dispatching client like here: https://github.com/theia-ide/theia/blob/37155c6a897b6ec57bb7765b207dd5389e297011/packages/messages/src/node/messages-backend-module.ts#L18

Reply:
Ideally, each client would receive its own instance or `SearchInWorkspaceServer`, because searches by one client are not related to searches of another client.  So we don't want the server send `onResult` and `onDone` to other clients than the one that made the request.  I'm not sure how to do that though.

Reply:
Ah, not making those singletons seems to work:

```diff
diff --git a/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts b/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts
index 88908ff..696fe67 100644
--- a/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts
+++ b/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts
@@ -17,7 +17,7 @@ import {
 import { RipgrepSearchInWorkspaceServer } from "./ripgrep-search-in-workspace-server";
 
 export default new ContainerModule(bind => {
-        bind(SearchInWorkspaceServer).to(RipgrepSearchInWorkspaceServer).inSingletonScope();
+        bind(SearchInWorkspaceServer).to(RipgrepSearchInWorkspaceServer);
         bind(ConnectionHandler).toDynamicValue(ctx =>
                 new JsonRpcConnectionHandler<SearchInWorkspaceClient>
                         ('/search-in-workspace', client => {
@@ -26,5 +26,5 @@ export default new ContainerModule(bind => {
                                 client.onDidCloseConnection(() => server.dispose());
                                 return server;
                         })
-        ).inSingletonScope();
+        );
 });

```